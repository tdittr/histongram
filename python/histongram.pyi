from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
