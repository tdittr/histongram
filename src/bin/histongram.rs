use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
