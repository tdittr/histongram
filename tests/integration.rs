use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::Histogram;

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a = Histogram::from_iter(APACHE.split_whitespace());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.split_whitespace());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::from_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
