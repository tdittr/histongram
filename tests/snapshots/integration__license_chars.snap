---
source: tests/integration.rs
expression: a_counts
---
[
  (' ', 2515),
  ('e', 833),
  ('t', 742),
  ('i', 729),
  ('o', 729),
  ('n', 615),
  ('r', 587),
  ('a', 518),
  ('s', 450),
  ('c', 308),
  ('h', 307),
  ('d', 283),
  ('l', 259),
  ('u', 244),
  ('
', 200),
  ('f', 178),
  ('m', 164),
  ('p', 156),
  ('y', 152),
  (',', 141),
  ('b', 138),
  ('g', 122),
  ('w', 94),
  ('v', 72),
  ('k', 65),
  ('.', 60),
  ('W', 52),
  ('L', 51),
  ('"', 40),
  ('I', 40),
  ('C', 37),
  ('N', 37),
  ('S', 36),
  ('Y', 36),
  ('D', 34),
  ('O', 33),
  ('T', 33),
  ('A', 26),
  ('E', 26),
  ('(', 23),
  (')', 23),
  ('x', 21),
  ('R', 19),
  ('-', 11),
  ('F', 10),
  ('U', 10),
  ('/', 9),
  ('j', 8),
  ('P', 7),
  ('0', 6),
  (';', 6),
  ('q', 6),
  ('2', 5),
  (':', 5),
  ('H', 5),
  ('B', 4),
  ('M', 4),
  ('G', 3),
  ('[', 3),
  (']', 3),
  ('1', 2),
  ('4', 2),
  ('5', 2),
  ('9', 2),
  ('K', 2),
  ('V', 2),
  ('z', 2),
  ('!', 1),
  ('%', 1),
  ('\'', 1),
  ('3', 1),
  ('6', 1),
  ('7', 1),
  ('8', 1),
  ('J', 1),
  ('X', 1),
]
